extensions/amp-analytics/0.1/events.js

```diff
@@ -1203,6 +1221,34 @@ export class VisibilityTracker extends EventTracker {
     });
   }
 
+  /**
+   * Returns a Promise indicating that we're ready to report the analytics,
+   * in the case of reportWhen: documentExit
+   * @return {!Promise}
+   * @private
+   */
+  createReportReadyPromiseForDocumentExit_() {
+    const deferred = new Deferred();
+    const root = this.root.getRoot();
+    let unloadListener, pageHideListener;
+    // Fired when document is being unloaded
+    root.addEventListener('unload', unloadListener = () => {
+      root.removeEventListener('unload', unloadListener);
+      deferred.resolve();
+    });
+    // Note: currently not supported on Opera Mini, nor IE<=10.
+    // Indicates traversing away from a session history item
+    // Safari on iOS will also fire it when switching tabs or switching to
```

Comment:
Thanks, those are super helpful.

More questions:
does it fire multiple times if user switch tabs back-and-forth? Also, does it fire when user click home button and put browser into background?

Reply:
I think this chart from https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/ explains it best:
![image](https://user-images.githubusercontent.com/2455047/47816643-14ac4800-dd2a-11e8-9afb-64fa9d0f65ce.png)

I am currently testing what happens when we have repeat:true with this...

Reply:
...but note that that chart is from 2015.